File: dlitscustom/dlitscustom/report/dlits_sales_partner_commission_report/dlits_sales_partner_commission_report.js
// Copyright (c) 2024, DLITS and contributors
// For license information, please see license.txt

frappe.query_reports["DLITS Sales Partner Commission Report"] = {
	"filters": [
		{
			"fieldname": "from_date",
			"label": __("From Date"),
			"fieldtype": "Date",
			"default": frappe.datetime.add_months(frappe.datetime.get_today(), -3),
			"reqd": 1
		},
		{
			"fieldname": "to_date",
			"label": __("To Date"),
			"fieldtype": "Date",
			"default": frappe.datetime.get_today(),
			"reqd": 1
		},
		{
			"fieldname": "sales_partner",
			"label": __("DLITS Sales Partner"),
			"fieldtype": "Link",
			"options": "DLITS Sales Partner"
		},
		{
			"fieldname": "commission_rate_min",
			"label": __("Min Commission Rate (%)"),
			"fieldtype": "Float"
		},
		{
			"fieldname": "commission_rate_max",
			"label": __("Max Commission Rate (%)"),
			"fieldtype": "Float"
		},
		{
			"fieldname": "include_disabled",
			"label": __("Include Disabled"),
			"fieldtype": "Check",
			"default": 0
		},
		{
			"fieldname": "show_all",
			"label": __("Show All (Including Zero Commission)"),
			"fieldtype": "Check",
			"default": 0
		}
	],
	
	"formatter": function(value, row, column, data, default_formatter) {
		value = default_formatter(value, row, column, data);
		
		if (column.fieldname == "payment_status" && data) {
			let color = "blue";
			if (data.payment_status == "Fully Paid") color = "green";
			else if (data.payment_status == "Unpaid") color = "red";
			else if (data.payment_status == "Partially Paid") color = "orange";
			else if (data.payment_status == "No Commission") color = "gray";
			
			value = `<span class="indicator ${color}">${data.payment_status}</span>`;
		}
		
		if (column.fieldname == "outstanding_commission" && data && data.outstanding_commission > 0) {
			value = `<span class="text-danger"><strong>${value}</strong></span>`;
		}
		
		if (column.fieldname == "commission_rate" && data && data.commission_rate) {
			value = `${data.commission_rate}%`;
		}
		
		return value;
	},
	
	"onload": function(report) {
		// Add custom buttons
		report.page.add_inner_button(__("Create Commission Payment"), function() {
			let selected_rows = report.datatable.rowmanager.getCheckedRows();
			if (selected_rows.length === 0) {
				frappe.msgprint(__("Please select at least one sales partner"));
				return;
			}
			
			let sales_partners = selected_rows.map(row => {
				let data = report.data[row];
				return {
					sales_partner: data.sales_partner,
					partner_name: data.partner_name,
					outstanding_commission: data.outstanding_commission,
					linked_supplier: data.linked_supplier
				};
			});
			
			show_commission_payment_dialog(sales_partners);
		});
		
		report.page.add_inner_button(__("Outstanding Commissions"), function() {
			frappe.call({
				method: "dlitscustom.utils.commission_management_dlits.get_outstanding_commissions",
				args: {
					from_date: report.get_filter_value("from_date"),
					to_date: report.get_filter_value("to_date")
				},
				callback: function(r) {
					if (r.message && r.message.length > 0) {
						show_outstanding_commissions_dialog(r.message);
					} else {
						frappe.msgprint(__("No outstanding commissions found"));
					}
				}
			});
		});
		
		report.page.add_inner_button(__("Commission Summary"), function() {
			let sales_partner = report.get_filter_value("sales_partner");
			if (!sales_partner) {
				frappe.msgprint(__("Please select a DLITS Sales Partner first"));
				return;
			}
			
			frappe.call({
				method: "dlitscustom.utils.commission_management_dlits.get_sales_partner_commission_summary",
				args: {
					sales_partner: sales_partner,
					from_date: report.get_filter_value("from_date"),
					to_date: report.get_filter_value("to_date")
				},
				callback: function(r) {
					if (r.message) {
						show_commission_summary_dialog(r.message);
					}
				}
			});
		});
		
		report.page.add_inner_button(__("Export Outstanding"), function() {
			let outstanding_data = report.data.filter(row => row.outstanding_commission > 0);
			if (outstanding_data.length === 0) {
				frappe.msgprint(__("No outstanding commissions to export"));
				return;
			}
			
			// Create CSV content
			let csv_content = "Sales Partner,Partner Name,Outstanding Commission,Linked Supplier\n";
			outstanding_data.forEach(row => {
				csv_content += `"${row.sales_partner}","${row.partner_name}","${row.outstanding_commission}","${row.linked_supplier || ''}"\n`;
			});
			
			// Download CSV
			let blob = new Blob([csv_content], { type: 'text/csv' });
			let url = window.URL.createObjectURL(blob);
			let a = document.createElement('a');
			a.href = url;
			a.download = `outstanding_commissions_${frappe.datetime.get_today()}.csv`;
			a.click();
			window.URL.revokeObjectURL(url);
		});
	}
};

function show_commission_payment_dialog(sales_partners) {
	let total_outstanding = sales_partners.reduce((sum, sp) => sum + (sp.outstanding_commission || 0), 0);
	
	let d = new frappe.ui.Dialog({
		title: __("Create Commission Payments"),
		fields: [
			{
				fieldname: "payment_date",
				label: __("Payment Date"),
				fieldtype: "Date",
				default: frappe.datetime.get_today(),
				reqd: 1
			},
			{
				fieldname: "mode_of_payment",
				label: __("Mode of Payment"),
				fieldtype: "Link",
				options: "Mode of Payment",
				reqd: 1
			},
			{
				fieldname: "total_amount",
				label: __("Total Outstanding Amount"),
				fieldtype: "Currency",
				read_only: 1,
				default: total_outstanding
			},
			{
				fieldname: "sales_partners_html",
				fieldtype: "HTML"
			}
		],
		primary_action_label: __("Create Payments"),
		primary_action: function(values) {
			let payments_data = sales_partners.map(sp => ({
				sales_partner: sp.sales_partner,
				amount: sp.outstanding_commission,
				reference_date: values.payment_date
			}));
			
			frappe.call({
				method: "dlitscustom.utils.commission_management_dlits.create_bulk_commission_payments",
				args: {
					sales_partners_data: payments_data
				},
				callback: function(r) {
					if (r.message) {
						let result = r.message;
						let message = `Created ${result.created_payments.length} payment(s)`;
						if (result.errors.length > 0) {
							message += `<br>Errors: ${result.errors.join('<br>')}`;
						}
						frappe.msgprint({
							title: __("Payment Creation Result"),
							message: message,
							indicator: result.errors.length > 0 ? "orange" : "green"
						});
						cur_page.page.page.refresh();
					}
				}
			});
			d.hide();
		}
	});
	
	// Show sales partners table
	let html = `<table class="table table-bordered">
		<thead>
			<tr>
				<th>Sales Partner</th>
				<th>Partner Name</th>
				<th>Outstanding Commission</th>
				<th>Linked Supplier</th>
			</tr>
		</thead>
		<tbody>`;
	
	sales_partners.forEach(sp => {
		html += `<tr>
			<td>${sp.sales_partner}</td>
			<td>${sp.partner_name}</td>
			<td>${format_currency(sp.outstanding_commission)}</td>
			<td>${sp.linked_supplier || 'Not Linked'}</td>
		</tr>`;
	});
	
	html += `</tbody></table>`;
	d.fields_dict.sales_partners_html.$wrapper.html(html);
	d.show();
}

function show_outstanding_commissions_dialog(outstanding_data) {
	let d = new frappe.ui.Dialog({
		title: __("Outstanding Commissions Summary"),
		size: "large",
		fields: [
			{
				fieldname: "outstanding_html",
				fieldtype: "HTML"
			}
		]
	});
	
	let total_outstanding = outstanding_data.reduce((sum, item) => sum + item.outstanding_commission, 0);
	
	let html = `<div class="row">
		<div class="col-md-12">
			<h4>Total Outstanding: ${format_currency(total_outstanding)}</h4>
			<table class="table table-bordered">
				<thead>
					<tr>
						<th>Sales Partner</th>
						<th>Partner Name</th>
						<th>Commission Rate</th>
						<th>Total Commission</th>
						<th>Paid Commission</th>
						<th>Outstanding Commission</th>
					</tr>
				</thead>
				<tbody>`;
	
	outstanding_data.forEach(item => {
		html += `<tr>
			<td><a href="/app/dlits-sales-partner/${item.sales_partner}">${item.sales_partner}</a></td>
			<td>${item.partner_name}</td>
			<td>${item.commission_rate}%</td>
			<td>${format_currency(item.total_commission)}</td>
			<td>${format_currency(item.paid_commission)}</td>
			<td><strong>${format_currency(item.outstanding_commission)}</strong></td>
		</tr>`;
	});
	
	html += `</tbody></table></div></div>`;
	d.fields_dict.outstanding_html.$wrapper.html(html);
	d.show();
}

function show_commission_summary_dialog(summary) {
	let d = new frappe.ui.Dialog({
		title: __("Commission Summary for {0}", [summary.sales_partner]),
		size: "large",
		fields: [
			{
				fieldname: "summary_html",
				fieldtype: "HTML"
			}
		]
	});
	
	let html = `<div class="row">
		<div class="col-md-6">
			<h5>Commission Overview</h5>
			<table class="table">
				<tr><td>Total Commission:</td><td><strong>${format_currency(summary.total_commission)}</strong></td></tr>
				<tr><td>Paid Commission:</td><td>${format_currency(summary.paid_commission)}</td></tr>
				<tr><td>Outstanding Commission:</td><td><strong class="text-danger">${format_currency(summary.outstanding_commission)}</strong></td></tr>
			</table>
		</div>
		<div class="col-md-6">
			<h5>Sales Overview</h5>
			<table class="table">
				<tr><td>Total Orders:</td><td>${summary.total_orders}</td></tr>
				<tr><td>Total Sales:</td><td>${format_currency(summary.total_sales)}</td></tr>
				<tr><td>Commission Rate:</td><td>${summary.commission_rate}%</td></tr>
				<tr><td>Linked Supplier:</td><td>${summary.linked_supplier || 'Not Linked'}</td></tr>
			</table>
		</div>
	</div>`;
	
	d.fields_dict.summary_html.$wrapper.html(html);
	d.show();
}

File: dlitscustom/dlitscustom/report/dlits_sales_partner_commission_report/dlits_sales_partner_commission_report.py
# Copyright (c) 2024, DLITS and contributors
# For license information, please see license.txt

import frappe
from frappe import _
from frappe.utils import flt, getdate, today
from dlitscustom.utils.commission_management_dlits import get_sales_partner_commission_summary

def execute(filters=None):
    columns = get_columns()
    data = get_data(filters)
    chart = get_chart_data(data, filters)
    return columns, data, None, chart

def get_columns():
    return [
        {
            "label": _("DLITS Sales Partner"),
            "fieldname": "sales_partner",
            "fieldtype": "Link",
            "options": "DLITS Sales Partner",
            "width": 150
        },
        {
            "label": _("Partner Name"),
            "fieldname": "partner_name",
            "fieldtype": "Data",
            "width": 150
        },
        {
            "label": _("Commission Rate (%)"),
            "fieldname": "commission_rate",
            "fieldtype": "Percent",
            "width": 120
        },
        {
            "label": _("Total Sales"),
            "fieldname": "total_sales",
            "fieldtype": "Currency",
            "width": 120
        },
        {
            "label": _("Total Orders"),
            "fieldname": "total_orders",
            "fieldtype": "Int",
            "width": 100
        },
        {
            "label": _("Total Commission"),
            "fieldname": "total_commission",
            "fieldtype": "Currency",
            "width": 130
        },
        {
            "label": _("Paid Commission"),
            "fieldname": "paid_commission",
            "fieldtype": "Currency",
            "width": 130
        },
        {
            "label": _("Outstanding Commission"),
            "fieldname": "outstanding_commission",
            "fieldtype": "Currency",
            "width": 150
        },
        {
            "label": _("Payment Status"),
            "fieldname": "payment_status",
            "fieldtype": "Data",
            "width": 120
        },
        {
            "label": _("Linked Supplier"),
            "fieldname": "linked_supplier",
            "fieldtype": "Data",
            "width": 150
        },
        {
            "label": _("Recent Sales Invoices"),
            "fieldname": "recent_invoices",
            "fieldtype": "Data",
            "width": 200
        },
        {
            "label": _("Recent Payments"),
            "fieldname": "recent_payments",
            "fieldtype": "Data",
            "width": 180
        },
        {
            "label": _("Last Transaction"),
            "fieldname": "last_transaction_date",
            "fieldtype": "Date",
            "width": 120
        }
    ]

def get_data(filters):
    conditions = get_conditions(filters)
    from_date = filters.get("from_date") or frappe.utils.add_months(today(), -3)
    to_date = filters.get("to_date") or today()
    
    # Get all DLITS sales partners
    sales_partners = frappe.db.sql(f"""
        SELECT
            sp.name as sales_partner,
            sp.partner_name,
            sp.commission_rate,
            sp.status,
            sp.supplier as linked_supplier
        FROM `tabDLITS Sales Partner` sp
        WHERE 1=1 {conditions}
        ORDER BY sp.partner_name
    """, as_dict=True)
    
    data = []
    
    for sp in sales_partners:
        # Get commission summary for this sales partner
        summary = get_sales_partner_commission_summary(
            sp.sales_partner, from_date, to_date
        )
        
        # Get the actual outstanding commission from DLITS Sales Partner record
        partner_record = frappe.get_doc("DLITS Sales Partner", sp.sales_partner)
        actual_outstanding = flt(partner_record.outstanding_commission or 0)
        actual_total_commission = flt(partner_record.total_commission_earned or 0)
        actual_paid_commission = flt(partner_record.total_commission_paid or 0)
        actual_total_sales = flt(partner_record.total_sales or 0)
        
        # Use the higher values between calculated and stored
        final_total_commission = max(summary['total_commission'], actual_total_commission)
        final_paid_commission = max(summary['paid_commission'], actual_paid_commission)
        final_outstanding_commission = max(summary['outstanding_commission'], actual_outstanding)
        final_total_sales = max(summary['total_sales'], actual_total_sales)
        
        # Get recent sales invoices (last 3)
        recent_invoices = frappe.db.sql("""
            SELECT name, posting_date, base_grand_total, dlits_commission_amount
            FROM `tabSales Invoice`
            WHERE dlits_sales_partner = %s AND docstatus = 1
            ORDER BY posting_date DESC
            LIMIT 3
        """, (sp.sales_partner,), as_dict=True)
        
        # Format recent invoices display
        recent_invoices_display = ""
        if recent_invoices:
            invoice_list = []
            for inv in recent_invoices:
                invoice_list.append(f"{inv.name} ({frappe.format(inv.dlits_commission_amount, {'fieldtype': 'Currency'})})")
            recent_invoices_display = ", ".join(invoice_list)
        else:
            recent_invoices_display = "No invoices found"
        
        # Get recent payments (last 3)
        recent_payments_display = "No payments found"
        if sp.linked_supplier:
            # Get supplier name for display
            supplier_name = frappe.db.get_value("Supplier", sp.linked_supplier, "supplier_name") or sp.linked_supplier
            
            recent_payments = frappe.db.sql("""
                SELECT name, reference_date, paid_amount
                FROM `tabPayment Entry`
                WHERE party_type = 'Supplier' AND party = %s AND docstatus = 1
                AND (remarks LIKE %s OR remarks LIKE %s)
                ORDER BY reference_date DESC
                LIMIT 3
            """, (sp.linked_supplier, f"%Commission%{sp.sales_partner}%", f"%commission%{sp.sales_partner}%"), as_dict=True)
            
            if recent_payments:
                payment_list = []
                for payment in recent_payments:
                    payment_list.append(f"{payment.name} ({frappe.format(payment.paid_amount, {'fieldtype': 'Currency'})})")
                recent_payments_display = ", ".join(payment_list)
        
        # Get last transaction date (only from Sales Invoices)
        last_transaction = frappe.db.sql("""
            SELECT MAX(posting_date) as last_date
            FROM `tabSales Invoice`
            WHERE dlits_sales_partner = %s AND docstatus = 1
        """, (sp.sales_partner,), as_dict=True)
        
        last_transaction_date = last_transaction[0].last_date if last_transaction and last_transaction[0].last_date else None
        
        # Determine payment status
        payment_status = get_payment_status(final_outstanding_commission, final_total_commission)
        
        # Get supplier display name
        supplier_display = "Not Linked"
        if sp.linked_supplier:
            supplier_name = frappe.db.get_value("Supplier", sp.linked_supplier, "supplier_name")
            supplier_display = supplier_name or sp.linked_supplier
        
        # Include if there's any commission (from transactions or stored) or if show_all is enabled
        if final_total_commission > 0 or final_outstanding_commission > 0 or filters.get('show_all'):
            data.append({
                "sales_partner": sp.sales_partner,
                "partner_name": sp.partner_name,
                "commission_rate": sp.commission_rate,
                "total_sales": final_total_sales,
                "total_orders": summary['total_orders'],
                "total_commission": final_total_commission,
                "paid_commission": final_paid_commission,
                "outstanding_commission": final_outstanding_commission,
                "payment_status": payment_status,
                "linked_supplier": supplier_display,
                "recent_invoices": recent_invoices_display,
                "recent_payments": recent_payments_display,
                "last_transaction_date": last_transaction_date,
                "disabled": sp.status != 'Active'
            })
    
    # Sort by outstanding commission (highest first)
    data.sort(key=lambda x: x['outstanding_commission'], reverse=True)
    
    return data

def get_conditions(filters):
    conditions = ""
    
    if not filters.get("include_disabled"):
        conditions += " AND sp.status = 'Active'"
    
    if filters.get("sales_partner"):
        conditions += f" AND sp.name = '{filters.get('sales_partner')}'"
    
    if filters.get("commission_rate_min"):
        conditions += f" AND sp.commission_rate >= {flt(filters.get('commission_rate_min'))}"
    
    if filters.get("commission_rate_max"):
        conditions += f" AND sp.commission_rate <= {flt(filters.get('commission_rate_max'))}"
    
    return conditions

def get_payment_status(outstanding, total):
    """Determine payment status based on outstanding vs total commission"""
    if total == 0:
        return "No Commission"
    elif outstanding == 0:
        return "Fully Paid"
    elif outstanding == total:
        return "Unpaid"
    else:
        return "Partially Paid"

def get_chart_data(data, filters):
    """Generate chart data for commission analysis"""
    
    if not data:
        return None
    
    # Payment Status Distribution
    status_counts = {}
    total_outstanding = 0
    total_paid = 0
    
    for row in data:
        status = row.get("payment_status", "Unknown")
        status_counts[status] = status_counts.get(status, 0) + 1
        total_outstanding += flt(row.get("outstanding_commission", 0))
        total_paid += flt(row.get("paid_commission", 0))
    
    return {
        "data": {
            "labels": ["Outstanding Commission", "Paid Commission"],
            "datasets": [
                {
                    "name": "Commission Status",
                    "values": [total_outstanding, total_paid]
                }
            ]
        },
        "type": "donut",
        "height": 300,
        "colors": ["#ff6b6b", "#51cf66"]
    }